server/recceiver/test_cfstore.py

```diff
@@ -0,0 +1,234 @@
+        deferred = yield self.cfstore.commit(TR10)
+        self.assertCommit([u'ch1', u'ch2', u'ch3', u'ch4', u'ch5', u'ch6', u'ch7'],
+                          {u'ch1': abbr(u'ch1', 'testhosta', 1111, 'Active'),
+                           u'ch2': abbr(u'ch2', 'testhosta', 1111, 'Active'),
+                           u'ch3': abbr(u'ch3', 'testhosta', 1111, 'Active'),
+                           u'ch4': abbr(u'ch4', 'testhosta', 1111, 'Active'),
+                           u'ch5': abbr(u'ch5', 'testhosta', 1111, 'Active'),
+                           u'ch6': abbr(u'ch6', 'testhosta', 1111, 'Active'),
+                           u'ch7': abbr(u'ch7', 'testhosta', 1111, 'Active')})
+
+    @defer.inlineCallbacks
+    def test_no_CFS(self):
+        self.cfclient = mock_client()
+        self.cfclient.connected = False
+        self.cfstore.client = self.cfclient
+        self.cfstore.startService()
+        TR1 = mock_TR()
+        TR1.src.host = 'testhosta'
+        TR1.src.port = 1111
+        TR1.addrec = {1: ('ch1', 'longin'), 5: ('ch5', 'longin'), 6: ('ch6', 'stringin'), 7: ('ch7', 'longout')}
+        threading._start_new_thread(self.simulate_reconnect, ())
+        deferred = yield self.cfstore.commit(TR1)
+        self.assertCommit([u'ch1', u'ch5', u'ch6', u'ch7'],
+                          {u'ch1': abbr(u'ch1', 'testhosta', 1111, 'Active'),
+                           u'ch5': abbr(u'ch5', 'testhosta', 1111, 'Active'),
+                           u'ch6': abbr(u'ch6', 'testhosta', 1111, 'Active'),
+                           u'ch7': abbr(u'ch7', 'testhosta', 1111, 'Active')})
+
+    @defer.inlineCallbacks
+    def test_set_fail(self):
+        self.cfclient = mock_client()
+        self.cfclient.connected = True
+        self.cfclient.fail_set = True
+        self.cfstore.client = self.cfclient
+        self.cfstore.startService()
+        TR1 = mock_TR()
+        TR1.src.host = 'testhosta'
+        TR1.src.port = 1111
+        TR1.addrec = {1: ('ch1', 'longin'), 5: ('ch5', 'longin'), 6: ('ch6', 'stringin'), 7: ('ch7', 'longout')}
+        threading._start_new_thread(self.simulate_reconnect, ())
+        deferred = yield self.cfstore.commit(TR1)
+        self.assertCommit([u'ch1', u'ch5', u'ch6', u'ch7'],
+                          {u'ch1': abbr(u'ch1', 'testhosta', 1111, 'Active'),
+                           u'ch5': abbr(u'ch5', 'testhosta', 1111, 'Active'),
+                           u'ch6': abbr(u'ch6', 'testhosta', 1111, 'Active'),
+                           u'ch7': abbr(u'ch7', 'testhosta', 1111, 'Active')})
+
+
+    def assertCommit(self, keys, dict):
+        self.assertSequenceEqual(self.cfstore.client.cf.keys(), keys)
+        self.assertDictEqual(self.cfstore.client.cf, dict)
+
+    def simulate_reconnect(self):
+        time.sleep(3)
+        self.cfclient.connected = True
+        self.cfclient.fail_set = False
```

Comment:
There is a potential race where the main thread sees ```not connected and not fail_set```, which presumable would confuse the test logic.